protocol/bc/validation.go

```diff
@@ -0,0 +1,142 @@
+package bc
+
+import (
+	"chain/errors"
+	"chain/protocol/vm"
+)
+
+type (
+	// ValidChecker can check its validity with respect to a given
+	// validation state.
+	validChecker interface {
+		// CheckValid checks the entry for validity w.r.t. the given
+		// validation state.
+		checkValid(*validationState) error
+	}
+
+	validationState struct {
+		// The ID of the blockchain
+		blockchainID Hash
+
+		// The enclosing transaction object
+		tx *TxEntries
+
+		// The ID of the nearest enclosing entry
+		entryID Hash
+
+		// The source position, for validating ValueSources
+		sourcePos uint64
+
+		// The destination position, for validating ValueDestinations
+		destPos uint64
+	}
+)
+
+var (
+	errBadTimeRange          = errors.New("bad time range")
+	errEmptyResults          = errors.New("transaction has no results")
+	errEntryType             = errors.New("invalid entry type")
+	errMismatchedAssetID     = errors.New("mismatched asset id")
+	errMismatchedBlock       = errors.New("mismatched block")
+	errMismatchedMerkleRoot  = errors.New("mismatched merkle root")
+	errMismatchedPosition    = errors.New("mismatched value source/dest positions")
+	errMismatchedReference   = errors.New("mismatched reference")
+	errMismatchedValue       = errors.New("mismatched value")
+	errMisorderedBlockHeight = errors.New("misordered block height")
+	errMisorderedBlockTime   = errors.New("misordered block time")
+	errNoPrevBlock           = errors.New("no previous block")
+	errNoSource              = errors.New("no source for value")
+	errNonemptyExtHash       = errors.New("non-empty extension hash")
+	errOverflow              = errors.New("arithmetic overflow/underflow")
+	errPosition              = errors.New("invalid source or destination position")
+	errTxVersion             = errors.New("invalid transaction version")
+	errUnbalanced            = errors.New("unbalanced")
+	errUntimelyTransaction   = errors.New("block timestamp outside transaction time range")
+	errVersionRegression     = errors.New("version regression")
+	errWrongBlockchain       = errors.New("wrong blockchain")
+	errZeroTime              = errors.New("timerange has one or two bounds set to zero")
+)
+
+// ValidateTx validates a transaction.
+func ValidateTx(tx *TxEntries, initialBlockID Hash) error {
+	vs := &validationState{
+		blockchainID: initialBlockID,
+	}
+	return tx.checkValid(vs)
+}
+
+// ValidateBlock validates a block and the transactions within.
+// The consensus program is executed only if prev is non-nil and
+// runProg is true.
+func ValidateBlock(b, prev *BlockEntries, initialBlockID Hash, runProg bool) error {
+	if prev == nil {
```

Comment:
Looks like if the caller provides a prev block on block 1, this function will attempt to validate against it. I think it would be better to ignore it (or maybe return an error) in that case.

I also think it'd be easier to notice that sort of thing — easier to read the code — if it were flatter:

```go
if b.Body.Height > 1 && prev == nil {
	return errors.WithDetailf(errNoPrevBlock, "height %d", b.Body.Height)
}
if b.Body.Height > 1 {
	// validate against prev block
}

// validate current block
```

(It's okay to repeat part of the test in another if statement. We often have an urge to factor those out but it is just as often more readable if we don't. Check out these two excerpts from [TEOPS](https://www.amazon.com/Elements-Programming-Style-2nd/dp/0070342075).)

<img width="605" alt="decisions1" src="https://cloud.githubusercontent.com/assets/4178/24437051/abb58da8-13f3-11e7-897d-750802f48131.png">

and

<img width="607" alt="decisions2" src="https://cloud.githubusercontent.com/assets/4178/24437053/b1b106a6-13f3-11e7-9599-846ef5ce275e.png">